iocore/net/quic/QUICConfig.cc

```diff
@@ -489,3 +462,223 @@ QUICConfig::release(QUICConfigParams *params)
+QUICCertConfig::reconfigure()
+{
+  SSLConfig::scoped_config params;
+  SSLCertLookup *lookup = new SSLCertLookup();
+
+  QUICMultiCertConfigLoader loader(params);
+  loader.load(lookup);
+
+  _config_id = configProcessor.set(_config_id, lookup);
+}
+
+SSLCertLookup *
+QUICCertConfig::acquire()
+{
+  return static_cast<SSLCertLookup *>(configProcessor.get(_config_id));
+}
+
+void
+QUICCertConfig::release(SSLCertLookup *lookup)
+{
+  configProcessor.release(_config_id, lookup);
+}
+
+//
+// QUICMultiCertConfigLoader
+//
+SSL_CTX *
+QUICMultiCertConfigLoader::default_server_ssl_ctx()
+{
+  return quic_new_ssl_ctx();
+}
+
+SSL_CTX *
+QUICMultiCertConfigLoader::init_server_ssl_ctx(std::vector<X509 *> &cert_list, const SSLMultiCertConfigParams *multi_cert_params)
+{
+  const SSLConfigParams *params = this->_params;
+
+  SSL_CTX *ctx = this->default_server_ssl_ctx();
+
+  if (multi_cert_params) {
+    if (multi_cert_params->dialog) {
+      // TODO: dialog support
+    }
+
+    if (multi_cert_params->cert) {
+      if (!SSLMultiCertConfigLoader::load_certs(ctx, cert_list, params, multi_cert_params)) {
+        goto fail;
+      }
+    }
+
+    // SSL_CTX_load_verify_locations() builds the cert chain from the
+    // serverCACertFilename if that is not nullptr.  Otherwise, it uses the hashed
+    // symlinks in serverCACertPath.
+    //
+    // if ssl_ca_name is NOT configured for this cert in ssl_multicert.config
+    //     AND
+    // if proxy.config.ssl.CA.cert.filename and proxy.config.ssl.CA.cert.path
+    //     are configured
+    //   pass that file as the chain (include all certs in that file)
+    // else if proxy.config.ssl.CA.cert.path is configured (and
+    //       proxy.config.ssl.CA.cert.filename is nullptr)
+    //   use the hashed symlinks in that directory to build the chain
+    if (!multi_cert_params->ca && params->serverCACertPath != nullptr) {
+      if ((!SSL_CTX_load_verify_locations(ctx, params->serverCACertFilename, params->serverCACertPath)) ||
+          (!SSL_CTX_set_default_verify_paths(ctx))) {
+        Error("invalid CA Certificate file or CA Certificate path");
+        goto fail;
+      }
+    }
+  }
+
+  if (params->clientCertLevel != 0) {
+    // TODO: client cert support
+  }
+
+  if (!SSLMultiCertConfigLoader::set_session_id_context(ctx, params, multi_cert_params)) {
+    goto fail;
+  }
+
+  if (params->server_tls13_cipher_suites != nullptr) {
+    if (!SSL_CTX_set_ciphersuites(ctx, params->server_tls13_cipher_suites)) {
+      Error("invalid tls server cipher suites in records.config");
+      goto fail;
+    }
+  }
+
+  if (params->server_groups_list != nullptr) {
+    if (!SSL_CTX_set1_groups_list(ctx, params->server_groups_list)) {
+      Error("invalid groups list for server in records.config");
+      goto fail;
+    }
+  }
+
+  // SSL_CTX_set_info_callback(ctx, ssl_callback_info);
+
+  SSL_CTX_set_alpn_select_cb(ctx, QUIC::ssl_select_next_protocol, nullptr);
+
+  if (SSLConfigParams::ssl_ocsp_enabled) {
```

Comment:
Looks like there are several dup code. It would be great if we could have virtual functions like `_setup_cipher_suites`, `_setup_alpn`, `_setup_ocsp`, etc. Adding those functions would make this function more readable.

If SSL and QUIC require exactly the same setup then we can remove the dup code, and if there's a difference we can see what part is different easily.

Reply:
I agree. That's one of motivations of #5040, the long long yak shaving.